File: ios/QingKeSchedule/Features/CourseDraft.swift
import Foundation

struct CourseScheduleDraft: Identifiable, Equatable {
    let id: String
    var dayOfWeek: Int
    var startPeriod: Int
    var endPeriod: Int
    var startWeek: Int
    var endWeek: Int
    var repeatRule: RepeatRule
    var classroom: String
}

enum CourseSaveEvaluation: Equatable {
    case invalid([ScheduleValidationIssue])
    case conflicting([ScheduleConflictDTO])
    case ready
}

struct CourseDraft {
    var id: String
    var name: String
    var teacher: String
    var color: String
    var schedules: [CourseScheduleDraft]

    private let baseline: CourseDTO

    init(
        course: CourseDTO? = nil,
        semester: SemesterDTO,
        now: Date = Date(),
        calendar: Calendar = ScheduleRules.gregorianCalendar()
    ) {
        let firstPeriod = semester.periods.first?.number ?? 1
        let sundayBasedWeekday = calendar.component(.weekday, from: now)
        let currentDayOfWeek = sundayBasedWeekday == 1 ? 7 : sundayBasedWeekday - 1

        if let course {
            id = course.id
            name = course.name
            teacher = course.teacher
            color = course.color
            schedules = course.schedules.map(Self.makeScheduleDraft)
            baseline = course
        } else {
            let schedule = CourseScheduleDraft(
                id: UUID().uuidString,
                dayOfWeek: currentDayOfWeek,
                startPeriod: firstPeriod,
                endPeriod: firstPeriod,
                startWeek: 1,
                endWeek: semester.totalWeeks,
                repeatRule: .every,
                classroom: ""
            )
            let courseID = UUID().uuidString
            id = courseID
            name = ""
            teacher = ""
            color = "#287B74"
            schedules = [schedule]
            baseline = CourseDTO(
                id: courseID,
                name: "",
                teacher: "",
                color: "#287B74",
                schedules: [Self.makeScheduleDTO(schedule)]
            )
        }
    }

    var isDirty: Bool { course() != baseline }

    func course() -> CourseDTO {
        CourseDTO(
            id: id,
            name: name.trimmingCharacters(in: .whitespacesAndNewlines),
            teacher: teacher.trimmingCharacters(in: .whitespacesAndNewlines),
            color: color,
            schedules: schedules.map(Self.makeScheduleDTO)
        )
    }

    func validationIssues(
        semester: SemesterDTO,
        calendar: Calendar = ScheduleRules.gregorianCalendar()
    ) -> [ScheduleValidationIssue] {
        ScheduleValidator.validate(
            ScheduleDataDTO(
                semester: semester,
                courses: [course()],
                updatedAt: "1970-01-01T00:00:00.000Z"
            ),
            calendar: calendar
        )
    }

    func evaluateSave(
        semester: SemesterDTO,
        existingCourses: [CourseDTO],
        calendar: Calendar = ScheduleRules.gregorianCalendar()
    ) -> CourseSaveEvaluation {
        let issues = validationIssues(semester: semester, calendar: calendar)
        guard issues.isEmpty else { return .invalid(issues) }

        let conflicts = ScheduleRules.conflicts(
            for: course(),
            against: existingCourses
        )
        return conflicts.isEmpty ? .ready : .conflicting(conflicts)
    }

    mutating func addSchedule(copying source: CourseScheduleDraft? = nil) {
        let template = source ?? schedules.last
        schedules.append(
            CourseScheduleDraft(
                id: UUID().uuidString,
                dayOfWeek: template?.dayOfWeek ?? 1,
                startPeriod: template?.startPeriod ?? 1,
                endPeriod: template?.endPeriod ?? 1,
                startWeek: template?.startWeek ?? 1,
                endWeek: template?.endWeek ?? 1,
                repeatRule: template?.repeatRule ?? .every,
                classroom: template?.classroom ?? ""
            )
        )
    }

    mutating func removeSchedule(id: String) {
        guard schedules.count > 1 else { return }
        schedules.removeAll { $0.id == id }
    }

    private static func makeScheduleDraft(_ schedule: CourseScheduleDTO) -> CourseScheduleDraft {
        CourseScheduleDraft(
            id: schedule.id,
            dayOfWeek: schedule.dayOfWeek,
            startPeriod: schedule.startPeriod,
            endPeriod: schedule.endPeriod,
            startWeek: schedule.startWeek,
            endWeek: schedule.endWeek,
            repeatRule: schedule.repeat,
            classroom: schedule.classroom
        )
    }

    private static func makeScheduleDTO(_ schedule: CourseScheduleDraft) -> CourseScheduleDTO {
        CourseScheduleDTO(
            id: schedule.id,
            dayOfWeek: schedule.dayOfWeek,
            startPeriod: schedule.startPeriod,
            endPeriod: schedule.endPeriod,
            startWeek: schedule.startWeek,
            endWeek: schedule.endWeek,
            repeat: schedule.repeatRule,
            classroom: schedule.classroom.trimmingCharacters(in: .whitespacesAndNewlines)
        )
    }
}

File: Android/app/src/main/java/com/qingke/schedule/draft/CourseDraft.kt
package com.qingke.schedule.draft

import com.qingke.schedule.domain.Course
import com.qingke.schedule.domain.CourseSchedule
import com.qingke.schedule.domain.RepeatRule
import com.qingke.schedule.domain.ScheduleConflict
import com.qingke.schedule.domain.ScheduleData
import com.qingke.schedule.domain.ScheduleValidator
import com.qingke.schedule.domain.Semester
import com.qingke.schedule.domain.ValidationIssue
import java.time.LocalDate
import java.util.UUID

data class CourseScheduleDraft(
    val id: String,
    var dayOfWeek: Int,
    var startPeriod: Int,
    var endPeriod: Int,
    var startWeek: Int,
    var endWeek: Int,
    var repeatRule: RepeatRule,
    var classroom: String,
) {
    fun toCourseSchedule(): CourseSchedule = CourseSchedule(
        id, dayOfWeek, startPeriod, endPeriod, startWeek, endWeek, repeatRule, classroom.trim(),
    )
}

sealed interface CourseSaveEvaluation {
    data class Invalid(val issues: List<ValidationIssue>) : CourseSaveEvaluation
    data class Conflicting(val conflicts: List<ScheduleConflict>) : CourseSaveEvaluation
    data object Ready : CourseSaveEvaluation
}

class CourseDraft private constructor(
    var id: String,
    var name: String,
    var teacher: String,
    var color: String,
    val schedules: MutableList<CourseScheduleDraft>,
    private val baseline: Course,
    private val idFactory: () -> String,
) {
    val isDirty: Boolean get() = course() != baseline

    fun course(): Course = Course(
        id = id,
        name = name.trim(),
        teacher = teacher.trim(),
        color = color,
        schedules = schedules.map { it.toCourseSchedule() },
    )

    fun validationIssues(semester: Semester): List<ValidationIssue> = ScheduleValidator.validate(
        ScheduleData(1, semester, listOf(course()), "1970-01-01T00:00:00Z"),
    )

    fun evaluateSave(semester: Semester, existingCourses: List<Course>): CourseSaveEvaluation {
        val issues = validationIssues(semester)
        if (issues.isNotEmpty()) return CourseSaveEvaluation.Invalid(issues)
        val candidate = course()
        if (hasNewDuplicateSchedule(candidate)) {
            return CourseSaveEvaluation.Invalid(
                listOf(ValidationIssue("courses.0.schedules", "该上课安排已存在，请勿重复添加")),
            )
        }
        val conflicts = com.qingke.schedule.domain.ScheduleRules.conflicts(candidate, existingCourses)
        return if (conflicts.isEmpty()) CourseSaveEvaluation.Ready else CourseSaveEvaluation.Conflicting(conflicts)
    }

    fun addSchedule(copying: CourseScheduleDraft? = null) {
        val source = copying ?: schedules.lastOrNull()
        schedules += CourseScheduleDraft(
            idFactory(), source?.dayOfWeek ?: 1, source?.startPeriod ?: 1, source?.endPeriod ?: 1,
            source?.startWeek ?: 1, source?.endWeek ?: 1, source?.repeatRule ?: RepeatRule.EVERY,
            source?.classroom ?: "",
        )
    }

    fun removeSchedule(id: String) {
        if (schedules.size > 1) {
            val index = schedules.indexOfFirst { it.id == id }
            if (index >= 0) schedules.removeAt(index)
        }
    }

    private fun hasNewDuplicateSchedule(candidate: Course): Boolean {
        val baselineCounts = baseline.schedules.groupingBy(::feature).eachCount()
        return candidate.schedules.groupingBy(::feature).eachCount().any { (key, count) ->
            count > 1 && count > (baselineCounts[key] ?: 0)
        }
    }

    private data class Feature(
        val day: Int, val startPeriod: Int, val endPeriod: Int, val startWeek: Int, val endWeek: Int,
        val repeat: RepeatRule, val classroom: String,
    )

    private fun feature(schedule: CourseSchedule) = Feature(
        schedule.dayOfWeek, schedule.startPeriod, schedule.endPeriod, schedule.startWeek, schedule.endWeek,
        schedule.repeatRule, schedule.classroom.trim(),
    )

    companion object {
        fun create(
            semester: Semester,
            now: LocalDate,
            idFactory: () -> String = { UUID.randomUUID().toString() },
        ): CourseDraft {
            val courseId = idFactory()
            val draftSchedule = defaultSchedule(semester, now, idFactory())
            val baseline = Course(courseId, "", "", "#287B74", listOf(draftSchedule.toCourseSchedule()))
            return CourseDraft(courseId, "", "", "#287B74", mutableListOf(draftSchedule), baseline, idFactory)
        }

        fun edit(
            course: Course,
            semester: Semester,
            now: LocalDate,
            appendSchedule: Boolean = false,
            idFactory: () -> String = { UUID.randomUUID().toString() },
        ): CourseDraft {
            val drafts = course.schedules.map { schedule ->
                CourseScheduleDraft(schedule.id, schedule.dayOfWeek, schedule.startPeriod, schedule.endPeriod,
                    schedule.startWeek, schedule.endWeek, schedule.repeatRule, schedule.classroom)
            }.toMutableList()
            if (appendSchedule) drafts += defaultSchedule(semester, now, idFactory())
            val normalizedBaseline = course.copy(
                name = course.name.trim(), teacher = course.teacher.trim(),
                schedules = course.schedules.map { it.copy(classroom = it.classroom.trim()) },
            )
            return CourseDraft(course.id, course.name, course.teacher, course.color, drafts, normalizedBaseline, idFactory)
        }

        private fun defaultSchedule(semester: Semester, now: LocalDate, id: String) = CourseScheduleDraft(
            id, now.dayOfWeek.value, semester.periods.firstOrNull()?.number ?: 1,
            semester.periods.firstOrNull()?.number ?: 1, 1, semester.totalWeeks, RepeatRule.EVERY, "",
        )
    }
}
